### PR TITLE
To solve #41

### PR DESCRIPTION
**Pull request by guolinke**
To solve #41

### CODE REVIEW

File: include/LightGBM/utils/text_reader.h [hunk 1]
```diff
@@ -22,9 +23,41 @@ class TextReader {
...
+  TextReader(const char* filename, bool is_skip_first_line):
+    filename_(filename), is_skip_first_line_(is_skip_first_line){
+    if (is_skip_first_line_) {
+      FILE* file;
+#ifdef _MSC_VER
+      fopen_s(&file, filename, "r");
+#else
+      file = fopen(filename, "r");
+#endif
+      std::stringstream ss;
```
Comment (xuehui1991) [inline include/LightGBM/utils/text_reader.h]: I think the naming is a little weird.

File: src/io/dataset.cpp [hunk 2]
```diff
@@ -28,13 +29,111 @@ Dataset::Dataset(const char* data_filename, const char* init_score_filename,
...
+        std::string name = io_config.label_column.substr(name_prefix.size());
+        if (name2idx.count(name) > 0) {
+          label_idx_ = name2idx[name];
+          Log::Info("use %s column as label", name.c_str());
+        } else {
+          Log::Fatal("cannot find label column: %s in data file", name.c_str());
+        }
+      } else {
+        Common::Atoi(io_config.label_column.c_str(), &label_idx_);
+        Log::Info("use %d-th column as label", label_idx_);
```
Comment (xuehui1991) [inline src/io/dataset.cpp]: I wonder the range of label_idx_. Does it have any posiible to ...

File: src/io/dataset.cpp [hunk 3]
```diff
@@ -190,26 +289,43 @@ void Dataset::ConstructBinMappers(int rank, int num_machines, const std::vector<
   // -1 means doesn't use this feature
   used_feature_map_ = std::vector<int>(sample_values.size(), -1);
   num_total_features_ = static_cast<int>(sample_values.size());
+
+  // fill feature_names_ if not header
+  if (feature_names_.size() <= 0) {
+    for (int i = 0; i < num_total_features_; i++) {
```
Comment (xuehui1991) [inline src/io/dataset.cpp]: ++i?

File: src/io/dataset.cpp [hunk 4]
```diff
@@ -276,6 +397,13 @@ void Dataset::ConstructBinMappers(int rank, int num_machines, const std::vector<
 
 
 void Dataset::LoadTrainData(int rank, int num_machines, bool is_pre_partition, bool use_two_round_loading) {
+  // don't support query id in data file when training parallel
```
Comment (xuehui1991) [inline src/io/dataset.cpp]: // don't support query id in data file when training in parallel?

File: src/io/metadata.cpp [hunk 5]
```diff
@@ -36,12 +36,31 @@ Metadata::~Metadata() {
...
+      delete[] weights_; 
+    }
+    weights_ = new float[num_data_];
+    num_weights_ = num_data_;
+    memset(weights_, 0, sizeof(float) * num_data_);
+  }
+  if (query_idx >= 0) {
+    if (query_boundaries_ != nullptr) {
+      Log::Info("using query id in data file, and ignore additional query file");
+      delete[] query_boundaries_;
```
Comment (xuehui1991) [inline src/io/metadata.cpp]: query_boundaries_ = nullprt ?

File: src/io/metadata.cpp [hunk 6]
```diff
@@ -36,12 +36,31 @@ Metadata::~Metadata() {
...
+    }
+    if (query_weights_ != nullptr) { delete[] query_weights_; }
```
Comment (xuehui1991) [inline src/io/metadata.cpp]: query_weights_ = nullptr?

File: src/io/parser.cpp [hunk 7]
```diff
@@ -76,43 +81,47 @@ Parser* Parser::CreateParser(const char* filename, int num_features, bool* has_l
   GetStatistic(line1.c_str(), &comma_cnt, &tab_cnt, &colon_cnt);
```
Comment (xuehui1991) [inline src/io/parser.cpp]: why comma_cnt1? why line1?

File: src/io/parser.cpp [hunk 8]
```diff
@@ -76,43 +81,47 @@ Parser* Parser::CreateParser(const char* filename, int num_features, bool* has_l
...
   if (line2.size() == 0) {
     // if only have one line on file
     if (colon_cnt > 0) {
-      ret =  new LibSVMParser();
-      if (num_features > 0 && has_label != nullptr) {
-        *has_label = CheckHasLabelForLibsvm(line1);
+      if (num_features > 0) {
+        has_label = CheckHasLabelForLibsvm(line1);
       }
+      ret = new LibSVMParser(has_label ? label_idx : -1);
```
Comment (xuehui1991) [inline src/io/parser.cpp]: can we take this part "has_label ? label_idx : -1" out ?

File: src/io/parser.cpp [hunk 9]
```diff
@@ -76,43 +81,47 @@ Parser* Parser::CreateParser(const char* filename, int num_features, bool* has_l
...
     } else if (tab_cnt > 0) {
-      ret = new TSVParser();
-      if (num_features > 0 && has_label != nullptr) {
-        *has_label = CheckHasLabelForTSV(line1, num_features);
+      if (num_features > 0 ) {
+        has_label = CheckHasLabelForTSV(line1, num_features);
       }
+      ret = new TSVParser(has_label ? label_idx : -1);
     } else if (comma_cnt > 0) {
```
Comment (xuehui1991) [inline src/io/parser.cpp]: too many "else if"....

Comment (xuehui1991) [inline src/io/parser.cpp]: too many "else if"...

File: src/io/dataset.cpp (hunk 2 above)
Comment (guolinke) [inline src/io/dataset.cpp]: thanks, will check this before loading data

**Pull request merged**